src/shared/dissect-image.c

```diff
@@ -1661,6 +1693,30 @@ int dissect_image_and_warn(
         }
 }
 
+bool dissected_image_can_do_verity(const DissectedImage *image, unsigned partition_number) {
+        int k;
+
+        if (image->can_verity)
+                return true;
+        k = PARTITION_VERITY_OF(partition_number);
+        if (k >= 0)
+                return true;
+        if (dissected_image_has_verity(image, partition_number))
+                return true;
+        return false;
+}
+
+bool dissected_image_has_verity(const DissectedImage *image, unsigned partition_number) {
+        int k;
+
+        if (image->verity)
+                return true;
+        k = PARTITION_VERITY_OF(partition_number);
+        if (k >= 0 && image->partitions[k].found)
+                return true;
+        return false;
+}
```

Comment:
actually, this isn't right. So previously "systemd-dissect" would show whether a *specific* partition has verity. This of course would only ever be true for the root partition, but it was supposed to be an inherent per-partition check. The `verity` and `can_verity` booleans in the DissectedImage structure indicated globally whether verity was available, i.e. whether verity is used at all, by any of the partitions.

With your change an image that consists of a GPT partition table with a root partition, a verity partition for it, pluse five others, would be shown as having verity for each one of these partitions, but that's simply incorrect: only the root partition actually has it.

I figure you should introduce a new, seperate field in DissectedImage called "external_verity" or so that encodes whether external verity is available...

and then only have:

```c
bool dissected_image_partition_has_verity(const DissectedImage *image, int p) {
        assert(image);

        if (p == PARTITION_ROOT && image->external_verity) /* Externally configured verity applies to the root partition only */
                return true;

        k = PARTITION_VERITY_OF(p);
        if (k < 0) /* This partition type has no verity concept */
                return false;

        return image->partitions[k].found;
}
```

Note that getting this right might not appear to matter much right now. However, I hope to add some special support for verity enabled /usr partitions (that are independent of the root partitions, so that one can combine a writable root with a verity protected /usr), and then this really starts to matter, since suddenly other partition types than just the root partition can have verity sensibly

Reply:
Interesting use case, hadn't thought of it, thanks for mentioning it.
Adding a new flag looks a bit awkward though - I'd like to propose another alternative first: the global has_verity is used only if the image has a single partition. Simple and self-contained, and should work with multiple verity partitions.
If you'd still prefer to have a separate flag instead, let me know and I can work on that instead.